pkg/controller/build/build_controller.go

```diff
@@ -327,428 +338,452 @@ func (ctrl *Controller) Run(parentCtx context.Context, workers int) {
+				ctrl.syncAvailableStatus(machineOSBuild)
+				return err
+			}
+			return nil
+		}
 	}
+	return ctrl.syncAvailableStatus(machineOSBuild)
+}
 
-	ps := newPoolState(pool)
+func (ctrl *Controller) updateMachineConfigPool(old, cur interface{}) {
+	oldPool := old.(*mcfgv1.MachineConfigPool).DeepCopy()
+	curPool := cur.(*mcfgv1.MachineConfigPool).DeepCopy()
+	klog.V(4).Infof("Updating MachineConfigPool %s", oldPool.Name)
+
+	moscOld, mosbOld, _ := ctrl.getConfigAndBuildForPool(oldPool)
+	moscNew, mosbNew, _ := ctrl.getConfigAndBuildForPool(curPool)
 
-	// Not a layered pool, so stop here.
-	if !ps.IsLayered() {
-		klog.V(4).Infof("MachineConfigPool %s is not opted-in for layering, ignoring", pool.Name)
-		return nil
+	doABuild, err := ctrlcommon.BuildDueToPoolChange(ctrl.imageBuilder, oldPool, curPool, moscNew, mosbNew)
+	if err != nil {
+		klog.Errorln(err)
+		ctrl.handleErr(err, curPool.Name)
+		return
 	}
 
+	//reBuild := shouldWeRebuild(oldPool, curPool)
+
 	switch {
-	case ps.IsInterrupted():
-		klog.V(4).Infof("MachineConfigPool %s is build interrupted, requeueing", pool.Name)
-		ctrl.enqueueMachineConfigPool(pool)
-		return nil
-	case ps.IsDegraded():
-		klog.V(4).Infof("MachineConfigPool %s is degraded, requeueing", pool.Name)
-		ctrl.enqueueMachineConfigPool(pool)
-		return nil
-	case ps.IsRenderDegraded():
-		klog.V(4).Infof("MachineConfigPool %s is render degraded, requeueing", pool.Name)
-		ctrl.enqueueMachineConfigPool(pool)
-		return nil
-	case ps.IsBuildPending():
-		klog.V(4).Infof("MachineConfigPool %s is build pending", pool.Name)
-		return nil
-	case ps.IsBuilding():
-		klog.V(4).Infof("MachineConfigPool %s is building", pool.Name)
-		return nil
-	case ps.IsBuildSuccess():
-		klog.V(4).Infof("MachineConfigPool %s has successfully built", pool.Name)
-		return nil
-	default:
-		shouldBuild, err := shouldWeDoABuild(ctrl.imageBuilder, pool, pool)
+	// We've transitioned from a layered pool to a non-layered pool.
+	case ctrlcommon.IsLayeredPool(moscOld, mosbOld) && !ctrlcommon.IsLayeredPool(moscNew, mosbNew):
+		klog.V(4).Infof("MachineConfigPool %s has opted out of layering", curPool.Name)
+		if err := ctrl.finalizeOptOut(moscNew, mosbNew, curPool); err != nil {
+			klog.Errorln(err)
+			ctrl.handleErr(err, curPool.Name)
+			return
+		}
+	// We need to do a build.
+	case doABuild:
+		klog.V(4).Infof("MachineConfigPool %s has changed, requiring a build", curPool.Name)
+		var status *mcfgv1alpha1.MachineOSBuildStatus
+		mosbNew, status, err = ctrl.CreateBuildFromConfig(moscNew)
 		if err != nil {
-			return fmt.Errorf("could not determine if a build is required for MachineConfigPool %q: %w", pool.Name, err)
+			klog.Errorln(err)
+			ctrl.handleErr(err, curPool.Name)
+			return
 		}
+		mosbNew.Status = *status
 
-		if shouldBuild {
-			return ctrl.startBuildForMachineConfigPool(ps)
+		if err := ctrl.startBuildForMachineConfigPool(moscNew, mosbNew); err != nil {
+			ctrl.syncAvailableStatus(mosbNew)
```

Comment:
question: Yeah I think we have to sync the avail status because the mosb is not a success from the start and we need to report that.
suggestion: Makes sense! I have replaced the original error reporting with aggregated errors to report both errors together.